### DISCUSSION TITLE
Maintain presto function list

### DISCUSSION THREAD

**Discussion by marin-ma** [apache/gluten#4759] [General]
#### Current solution in Gluten
Currently, functions under `velox/functions/prestosql` and `velox/functions/sparksql` are registered without prefix in Gluten
https://github.com/oap-project/gluten/blob/25223aab8ece0c420fac35c0e370c6305884080c/cpp/velox/operators/functions/RegistrationAllFunctions.cc#L66-L79

This solution can be risky in some scenarios:

1. Function A is implemented as simple function and registered in prestosql, while there is also a vector function implementation with same function signature in sparksql. As Velox maintain separated registration map for simple functions and vector functions and will first try to match the function in simple function registration then vector function registration, the vector function implementation in sparksql will never have a chance to be matched.
2. Function B is implemented as simple or vector function in both prestosql and sparksql, but the registered function signature in prestosql is not supported by Vanilla spark. An unsupported expression in vanilla spark can be successfully offloaded to the prestosql implementation without any information being exposed, which leads to inconsistency behavior with Vanilla spark.

#### Proposed change
We can maintain a presto function mapping in Gluten cpp during substrait->velox expresion conversion, and register all presto functions with "presto" prefix. To create the mapping, we need to find all supported functions that being offloaded to prestosql. For example, function `array_distinct` is using prestosql implementation, then add `array_distinct -> presto_array_distinct` to the mapping.
With the mapping, we can be sure that which functions goes to sparksql implementation, and which goes to prestosql.

**Comment by marin-ma** [comment]
cc: @zhouyuan @rui-mo @PHILO-HE

**Comment by philo-he** [comment]
Thanks for bringing up this discussion!
I remember Rui once proposed to include presto function header and explicitly register them for Spark on Velox side (cc @rui-mo). Then, there is no need to do any change in Gluten. If we found any semantic difference between spark & presto for a function, we can just implement a new one in sparksql folder and modify the registry. These changes are still on Velox side.

Not sure which is better. @FelixYBW, this issue was mentioned in the project meeting. Did you have any comment?
